src/theory/strings/inference_manager.cpp

```diff
@@ -349,20 +305,72 @@ void InferenceManager::doPendingFacts()
 
 void InferenceManager::doPendingLemmas()
 {
-  if (!d_state.isInConflict())
+  if (d_state.isInConflict())
+  {
+    // just clear the pending vectors, nothing else to do
+    d_pendingLem.clear();
+    d_pendingReqPhase.clear();
+    return;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  for (unsigned i = 0, psize = d_pendingLem.size(); i < psize; i++)
   {
-    for (const Node& lc : d_pendingLem)
+    InferInfo& ii = d_pendingLem[i];
+    Assert(!ii.isTrivial());
+    Assert(!ii.isConflict());
+    // get the explanation
+    Node eqExp;
+    if (options::stringRExplainLemmas())
     {
-      Trace("strings-pending") << "Process pending lemma : " << lc << std::endl;
-      ++(d_statistics.d_lemmasInfer);
-      d_out.lemma(lc);
+      eqExp = mkExplain(ii.d_ant, ii.d_antn);
     }
-    for (const std::pair<const Node, bool>& prp : d_pendingReqPhase)
+    else
     {
-      Trace("strings-pending") << "Require phase : " << prp.first
-                               << ", polarity = " << prp.second << std::endl;
-      d_out.requirePhase(prp.first, prp.second);
+      std::vector<Node> ev;
+      ev.insert(ev.end(), ii.d_ant.begin(), ii.d_ant.end());
+      ev.insert(ev.end(), ii.d_antn.begin(), ii.d_antn.end());
+      eqExp = utils::mkAnd(ev);
     }
+    // make the lemma node
+    Node lem = ii.d_conc;
+    if (eqExp != d_true)
+    {
+      lem = nm->mkNode(IMPLIES, eqExp, lem);
+    }
+    Trace("strings-pending") << "Process pending lemma : " << lem << std::endl;
+    Trace("strings-assert")
+        << "(assert " << lem << ") ; lemma " << ii.d_id << std::endl;
+    Trace("strings-lemma") << "Strings::Lemma: " << lem << " by " << ii.d_id
+                           << std::endl;
+    // only keep stats if we process it here
+    d_statistics.d_inferences << ii.d_id;
+    ++(d_statistics.d_lemmasInfer);
+    d_out.lemma(lem);
+
+    // Process the side effects of the inference info.
+    // [1] Register the new skolems from this inference. We register them here
+    // (lazily), since this is the moment when we have decided to use the
+    // inference at use_index that involves them.
+    for (const std::pair<const LengthStatus, std::vector<Node> >& sks :
+         ii.d_new_skolem)
+    {
+      for (const Node& n : sks.second)
+      {
+        d_termReg.registerTermAtomic(n, sks.first);
+      }
+    }
+    // [2] process the associated pending phase, add to map to process below
+    for (const std::pair<const Node, bool> pp : ii.d_pending_phase)
+    {
+      sendPhaseRequirement(pp.first, pp.second);
+    }
+  }
+  // process the pending require phase calls
+  for (const std::pair<const Node, bool>& prp : d_pendingReqPhase)
+  {
+    Trace("strings-pending") << "Require phase : " << prp.first
+                             << ", polarity = " << prp.second << std::endl;
+    d_out.requirePhase(prp.first, prp.second);
   }
```

Comment:
The issue with this is that someone may wish to set a phase without having a lemma.

I'm going to do a different simplification:  InferInfo does not contain d_pendingPhase.  It actually doesn't make sense to have it in InferInfo since:
(1) our inferences with pending phase requirements are restricted to lemma infer infos,
(2) pending lemmas and pending phase requirements are always processed at the same time. 
Thus, the called on sendInference / sendPhaseRequirement knows they are handled in sync.

Reply:
Yeah, that works, too. I am happy as long as we only have a single mechanism for setting phase requirements.